internal/serviceset/util.go

```diff
@@ -0,0 +1,293 @@
+		}
+		// we'll consider existing version can't be upgraded to the desired version
+		// in case existing version does not match version upgrade paths were computed for.
+		if upgradePath.Template != svc.Template {
+			return false
+		}
+		for _, upgradeList := range upgradePath.AvailableUpgrades {
+			if slices.Contains(upgradeList.Versions, desiredVersion) {
+				return true
+			}
+		}
+		return false
+	}
+	return res
+}
+
+// GetServiceSetWithOperation returns the ServiceSetOperation to perform and the ServiceSet object,
+// depending on the existence of the ServiceSet object and the services to deploy.
+func GetServiceSetWithOperation(
+	ctx context.Context,
+	c client.Client,
+	serviceSetObjectKey client.ObjectKey,
+	services []kcmv1.Service,
+	providerSpec kcmv1.StateManagementProviderConfig,
+) (*kcmv1.ServiceSet, kcmv1.ServiceSetOperation, error) {
+	l := ctrl.LoggerFrom(ctx)
+	serviceSet := new(kcmv1.ServiceSet)
+	err := c.Get(ctx, serviceSetObjectKey, serviceSet)
+	if client.IgnoreNotFound(err) != nil {
+		return nil, kcmv1.ServiceSetOperationNone, fmt.Errorf("failed to get ServiceSet %s: %w", serviceSetObjectKey, err)
+	}
+
+	switch {
+	case err != nil && len(services) == 0:
+		l.V(1).Info("No services to deploy, ServiceSet does not exist", "operation", kcmv1.ServiceSetOperationNone)
+		return nil, kcmv1.ServiceSetOperationNone, nil
+	case err != nil && len(services) > 0:
+		l.V(1).Info("Pending services to deploy, ServiceSet does not exist", "operation", kcmv1.ServiceSetOperationCreate)
+		serviceSet.SetName(serviceSetObjectKey.Name)
+		serviceSet.SetNamespace(serviceSetObjectKey.Namespace)
+		return serviceSet, kcmv1.ServiceSetOperationCreate, nil
+	case len(services) == 0:
+		l.V(1).Info("No services to deploy, ServiceSet exists", "operation", kcmv1.ServiceSetOperationDelete)
+		return serviceSet, kcmv1.ServiceSetOperationDelete, nil
+	case needsUpdate(serviceSet, providerSpec, services):
+		l.V(1).Info("Pending services to deploy, ServiceSet exists", "operation", kcmv1.ServiceSetOperationUpdate)
+		return serviceSet, kcmv1.ServiceSetOperationUpdate, nil
+	default:
+		l.V(1).Info("No actions required, ServiceSet exists", "operation", kcmv1.ServiceSetOperationNone)
+		return serviceSet, kcmv1.ServiceSetOperationNone, nil
+	}
+}
+
+// needsUpdate checks if the ServiceSet needs to be updated based on the ClusterDeployment spec.
+// It first compares the ServiceSet's provider configuration with the ClusterDeployment's service provider configuration.
+// Then it compares the ServiceSet's observed services' state with its desired state, and after that it compares
+// the ServiceSet's observed services' state with ClusterDeployment's desired services state.
+func needsUpdate(serviceSet *kcmv1.ServiceSet, providerSpec kcmv1.StateManagementProviderConfig, services []kcmv1.Service) bool {
+	// we'll need to update provider configuration if it was changed.
+	if !equality.Semantic.DeepEqual(providerSpec, serviceSet.Spec.Provider) {
```

Comment:
We don't need this check anymore ince `.spec.serviceSpec.provider.name` is immutable now.